pkg/controllers/work/apply_controller.go

```diff
@@ -629,6 +672,144 @@ func buildManifestAppliedCondition(err error, action applyAction, observedGenera
+			if err != nil {
+				klog.ErrorS(err, "failed to compute manifest hash", "manifest", obj.GetName(), "namespace", obj.GetNamespace())
+				return nil, err
+			}
+			lastModifiedConfig, err := computeModifiedConfiguration(obj)
+			if err != nil {
+				klog.ErrorS(err, "failed compute last applied config", "manifest", obj.GetName(), "namespace", obj.GetNamespace())
+				return nil, err
+			}
+			if err := r.createConfigMap(ctx, configMapName, owner, manifestHash, lastModifiedConfig); err != nil {
+				klog.ErrorS(err, "cannot create config map for object", "manifest", obj.GetName(), "namespace", obj.GetNamespace())
+				return nil, err
+			}
+			// get newly created config map recursively, doesn't look clean
+			newConfigMap, err := r.getConfigMap(ctx, obj, owner)
+			if err != nil {
+				return nil, err
+			}
+			return newConfigMap, nil
+		} else {
+			klog.ErrorS(err, "failed to retrieve config map for object", "manifest", obj.GetName(), "config map", configMapName)
+			return nil, err
+		}
+	}
+	return &configMap, nil
+}
+
+func (r *ApplyWorkReconciler) updateConfigMap(ctx context.Context, configMap *v1.ConfigMap, obj *unstructured.Unstructured) error {
+	updatedManifestHash, err := computeManifestHash(obj)
+	if err != nil {
+		// namespace for manifest can be empty
+		klog.ErrorS(err, "failed to compute new manifest hash", "manifest", obj.GetName(), "namespace", obj.GetNamespace())
+		return err
+	}
+	updatedLastModifiedConfig, err := computeModifiedConfiguration(obj)
+	if err != nil {
+		// namespace for manifest can be empty
+		klog.ErrorS(err, "failed to compute new last modified configuration", "manifest", obj.GetName(), "namespace", obj.GetNamespace())
+		return err
+	}
+	configMap.Data[manifestHashKey] = updatedManifestHash
+	configMap.Data[lastAppliedConfigKey] = updatedLastModifiedConfig
+	if err := r.spokeClient.Update(ctx, configMap); err != nil {
+		klog.ErrorS(err, "failed to update config map", "configMap", configMap.Name)
+		return err
+	}
+	return nil
+}
+
+// We can calculate the manifestHash & lastAppliedConfig again wiping out previous state or make sure the annotations in the last applied config doesn't cause issue in patch.
+func (r *ApplyWorkReconciler) migrateToConfigMap(ctx context.Context, gvr schema.GroupVersionResource, obj *unstructured.Unstructured, owner metav1.OwnerReference) error {
+	// we can also use obj.getAnnotations().
+	annots, err := metadataAccessor.Annotations(obj)
+	if err != nil {
+		return fmt.Errorf("cannot access metadata.annotations: %w", err)
+	}
+	if annots == nil {
+		return errors.New("manifest does not have any annotations")
+	}
+	// Assuming that manifest has both annotations on it
+	manifestHash, ok1 := annots[manifestHashAnnotation]
+	lastModifiedConfig, ok2 := annots[lastAppliedConfigAnnotation]
+	if ok1 || ok2 {
+		delete(annots, manifestHashAnnotation)
+		delete(annots, lastAppliedConfigAnnotation)
+		configMapName, err := r.getRandomConfigMapName(ctx)
```

Comment:
(See the comment above about refactoring the createConfigMap call; a little bit refactoring can help here)